dex/src/state.rs

```diff
@@ -1994,6 +1993,54 @@ pub(crate) mod account_parser {
             f(args)
         }
     }
+
+    pub struct CloseOpenOrdersArgs<'a, 'b: 'a> {
+        pub open_orders: &'a mut OpenOrders,
+        pub open_orders_acc: &'a AccountInfo<'b>,
+        pub dest_acc: &'a AccountInfo<'b>,
+    }
+
+    impl<'a, 'b: 'a> CloseOpenOrdersArgs<'a, 'b> {
+        pub fn with_parsed_args<T>(
+            program_id: &'a Pubkey,
+            accounts: &'a [AccountInfo<'b>],
+            f: impl FnOnce(CloseOpenOrdersArgs) -> DexResult<T>,
+        ) -> DexResult<T> {
+            // Parse accounts.
+            check_assert_eq!(accounts.len(), 4)?;
+            #[rustfmt::skip]
+            let &[
+                ref open_orders_acc,
+                ref owner_acc,
+                ref dest_acc,
+                ref market_acc,
+            ] = array_ref![accounts, 0, 4];
+
+            // Validate the accounts given are valid.
+            let owner = SignerAccount::new(owner_acc)?;
+            let market: RefMut<'a, MarketState> = MarketState::load(market_acc, program_id)?;
+            let mut open_orders =
+                market.load_orders_mut(open_orders_acc, Some(owner.inner()), program_id, None)?;
+
+            // Only accounts with no funds associated with them can be closed.
+            if open_orders.native_coin_total != 0 {
+                return Err(DexErrorCode::TooManyOpenOrders.into());
```

Comment:
This is the wrong error code.

Reply:
Moved the free slot bits check up and added logs instead of adding new error codes, as suggested.